models/biocro/R/run.biocro.R

```diff
@@ -51,111 +44,177 @@ run.biocro <- function(lat, lon, metpath, soil.nc = NULL, config = config, coppi
       day1 <- NULL
       dayn <- NULL
     }
+    if (!is.null(day1) && day1 > lubridate::yday(starti)) {
+      lubridate::yday(starti) <- day1
+    }
+    if (!is.null(dayn) && dayn < lubridate::yday(starti)) {
+      lubridate::yday(plant_date) <- dayn
+    }
+    WetDat <- WetDat[WetDat$doy >= lubridate::yday(starti) & WetDat$doy <= lubridate::yday(endi), ]
 
-    # BLETCHEROUS HACK: BioCro 0.94 starts the run by subsetting weather data
-    # to day1:dayn, but it assumes the data start on DOY 1 and contain
-    # (yearlength*(24/timestep)) lines. This means that in practice, day1 and
-    # dayn are treated as "day of file" not "day of year".
-    # BioCro *does* handle DOY correctly downstream of the subsetting, so here
-    # we check if the current BioCro has fixed this assumption.
-    # If not, rescale day1 and dayn to be relative to the start of the input.
-    #   Scaling is derived by inverting Biocro's day->index equations.
-    biocro_checks_doy <- tryCatch(
-      {m <- BioCro::BioGro(WetDat = matrix(c(0,10,0,0,0,0,0,0),nrow = 1),
-                   day1 = 10, dayn = 10, timestep = 24);
-      class(m) == "BioGro"},
-      error = function(e){FALSE})
-    if (!biocro_checks_doy && min(WetDat[,"doy"])>1) {
-      if (!is.null(day1)){
-        # Biocro calculates line number as `indes1 <- (day1 - 1) * 24`
-        indes1 <- Position(function(x)x==day1, WetDat[,"doy"])
-        day1 <- indes1/24 + 1
+    HarvestedYield <- 0
+
+    if (utils::packageVersion('BioCro') >= 1.0) {
```

Comment:
The conditional flow in this loop makes it very difficult to understand. Presumably, the two paths of the conditional perform the same fundamental calculations, and could be dependent on a class instead of control flow as follows.

Define the functions
results = biocro_0_wrapper(some arguments)
results = biocro_1_wrapper(some arguments)

and then write the loop as
```pseudocode
if (BioCro < version 1) {  # This if-else statement could and should be replaced by a lookup table if you continue to add more wrappers.
   loop_function = biocro_0_wrapper
} else {
   loop_function = biocro_1_wrapper
}
for (...) {
  do prep stuff
  tmp.result = loop_function(arguments)  # Now it's straightforward function call instead of branched control flow, and there's separation of responsibility between iteration and calculation.
}
```

The purpose of the loop is then clearer, and supporting new versions then requires only writing new wrappers instead of making an ever more complex for loop.

Reply:
that is a good idea. +1

Reply:
Fixed in a3d2e5a9f84f31847672a7bf934363fbc4794da2, but I pass it a bazillion arguments  to avoid changing any of the 0.9 logic, and I'm not sure I like the resulting kitchen-sink function signature. Further thoughts?

Reply:
Classes do this (in this specific case, a functor). Parameters specific to each model are passed upon initilization and each subclass returns a common interface. As a pseduocode example,
```
call_biocro_0.9 <- function(genus, ...) {  # ... means other parameters, not necessarily R's "the dots", but you could use them if they're convenient.
   function(WetData) {
      if (genus == 'Saccarhum') {
        tmp.result = CaneGro(...)
      } else if (genus == 'Salix') {
        tmp.result = willowGro(...)
      }
      return (tmp.result)
   }  # End of returned function.
}

call_biogro_1.0 <- function(...) {
   function(WetData) {  # This function has the same signature and return value as the one returned by call_biogro_0.9
      ...
   return (tmp.result)
   }
}

# In run.biocro
  if (utils::packageVersion('BioCro') >= 1.0) {
    caller_fn <- call_biocro_1(...)
  } else {
    caller_fn <- call_biocro_0.9(genus, ...)
  }
```

 BioCro 0.9 makes that someone annoying because of the way it picks start and end times, so you'd have subset WetDat before calling (in the same way you do for v1.0) and have the class return a function that sets day1 and dayn to the first and last days of the WetDat parameter, so you don't have to pass any time variables.

Reply:
Come to think of it... I already *do* subset WetDat before calling! So this is actually pretty straightforward to do.